journal.py

```diff
@@ -0,0 +1,233 @@
+
+    @classmethod
+    def write(cls, title=None, text=None, session=None):
+        if session is None:
+            session = DBSession
+        instance = cls(title=title, text=text)
+        session.add(instance)
+        return instance
+
+    id = sa.Column(sa.Integer, primary_key=True, autoincrement=True)
+    title = sa.Column(sa.Unicode(255), nullable=False)
+    created = sa.Column(
+        sa.DateTime, nullable=False, default=datetime.datetime.utcnow
+    )
+    text = sa.Column(sa.UnicodeText, nullable=False)
+
+    def __repr__(self):
+        return "<Entry(title='{}', creation_date='{}')>".format(
+            self.title, self.creation_date
+        )
+
+"""
+make a module-level constant for the connection URI
+(you'll need it elsewhere):
+"""
+DATABASE_URL = os.environ.get(
+    'DATABASE_URL',
+    'postgresql://tanner@localhost:5432/learning-journal',
+)
+
+
+def init_db():
+    engine = sa.create_engine(DATABASE_URL, echo=True)
+    Base.metadata.create_all(engine)
+
+
+def do_login(request):
+    username = request.params.get('username', None)
+    password = request.params.get('password', None)
+    if not (username and password):
+        raise ValueError('both username and password are required')
+
+    settings = request.registry.settings
+    # you can always get hold of application settings with
+    # `request.registry.settings`
+    manager = BCRYPTPasswordManager()
+    if username == settings.get('auth.username', ''):
+        hashed = settings.get('auth.password', '')
+        return manager.check(hashed, password)
+    return False
+
+
+# from pyramid.httpexceptions import HTTPNotFound
+
+
+@view_config(route_name='other', renderer='string')
+def other(request):
+    return request.matchdict
+
+
+@view_config(route_name='home', renderer='templates/list.jinja2')
+def list_view(request):
+    entries = Entry.all()
+    return {'entries': entries}
+
+
+@view_config(route_name='new_entry', renderer='templates/new_entry.jinja2')
+def new_entry(request):
+    return {}
+    # Turns out you need return *something*, even if that something
+    # is just an empty dict. The renderer is expecting a dict, but it
+    # doesn't need to contain anything useful.
+
+
+"""
+    Below:
+
+    Add a view function that will:
+
+        Pass values from the 'request' to our 'Entry.write()' method
+
+        Handle any exceptions raised by 'Entry.write()' appropriately,
+        returning a useful HTTP response
+
+        Send the viewer back to the home page if the entry was
+        successfully written
+
+    (We'll also need to configure a 'route' that will connect to
+    this new 'view function' --> 'config.add_route('add', '/add')' )
+"""
+
+
+@view_config(route_name='add', request_method='POST')
+def add_entry(request):
+    title = request.params.get('title')
+    text = request.params.get('text')
+    Entry.write(title=title, text=text)
+    return HTTPFound(request.route_url('home'))
```

Comment:
I'd love to see you work on combining `add_entry` and `new_entry` into a single view function.